librz/bin/format/le/le.c

```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
 	}
 }
 
-static char *le_read_nonnull_str_at(RzBuffer *buf, ut64 *offset) {
-	ut8 size;
-	if (!rz_buf_read8_at(buf, *offset, &size)) {
-		return NULL;
+static bool le_read_len_str_offset(RzBuffer *buf, ut64 *offset, char **out) {
+	*out = NULL;
+	ut8 len;
+	if (!rz_buf_read8_offset(buf, offset, &len)) {
+		return false;
+	}
+
+	if (!len) {
+		return true; // success yet *out == NULL, this is why the return value is bool
+	}
+
+	ut8 *str = calloc((size_t)len + 1, sizeof(char));
+	if (!str) {
+		return false;
+	}
+	if (!rz_buf_read_offset(buf, offset, str, len)) {
+		free(str);
+		return false;
+	}
+	for (ut8 *s = str; s != str + len; ++s) {
+		// non-ascii characters should not appear here
+		if (*s == 0 || *s > 127) {
+			free(str);
+			return false;
+		}
+	}
+	*out = (char *)str;
+	return true;
+}
+
+static ut32 le_reloc_target_offset(ut32 i) {
+	// TODO supposedly i860 / mips binaries exist, will 4 byte alignment suffice?
+	return i * 4;
+}
+
+static ut32 le_reloc_target_vaddr(rz_bin_le_obj_t *bin, ut32 i) {
+	return bin->reloc_target_map_base + le_reloc_target_offset(i);
+}
+
+static ut32 le_reloc_targets_vfile_size(rz_bin_le_obj_t *bin) {
+	return le_reloc_target_offset(bin->reloc_targets_count);
+}
+
+static ut32 le_obj_perm(LE_object *obj) {
+	ut32 perm = 0;
+	perm |= obj->flags & O_READABLE ? RZ_PERM_R : 0;
+	perm |= obj->flags & O_WRITABLE ? RZ_PERM_W : 0;
+	perm |= obj->flags & O_EXECUTABLE ? RZ_PERM_X : 0;
+	return perm;
+}
+
+static ut32 le_vaddr_to_paddr(rz_bin_le_obj_t *bin, ut32 vaddr) {
+	LE_map *m;
+	rz_vector_foreach(bin->le_maps, m) {
+		if (m->vaddr <= vaddr && vaddr <= m->vaddr + m->vsize) {
+			if (vaddr > m->vaddr + m->size) {
+				return 0;
+			} else {
+				return m->paddr + (vaddr - m->vaddr);
+			}
+		}
 	}
+	return 0;
+}
+
+static void le_import_free(LE_import *imp) {
+	if (imp) {
+		free(imp->proc_name);
+		free(imp);
+	}
+}
+
+static ut32 le_import_hash(LE_import *imp) {
+	ut32 ho = ((imp->mod_ord + 1) * (imp->proc_ord + 1) + 12345) * 1103515245;
+	return sdb_hash(imp->proc_name) ^ ho;
```

Comment:
A random hash. I couldn't find hash combining integers in rizin, and just slapped something this one in.

Reply:
awesome! just as i thought.
Instead of using multiplication, use xor.
like `ut32 ho = ((imp->mod_ord + 1) * (imp->proc_ord + 1) + 12345) ^ 1103515245;`

Reply:
because if you multiply by big number you can overflow the variable.

Reply:
But that is the intention, it is supposed to wrap around and keep the remainder modulo 2^32. I took generator from here https://en.wikipedia.org/wiki/Linear_congruential_generator#Parameters_in_common_use see line 5 of the table.